extensions/notification/NotificationQueuePopulator.js

```diff
@@ -150,132 +85,112 @@ class NotificationQueuePopulator extends QueuePopulatorExtension {
         return undefined;
     }
 
+    /**
+     * Extract base key from versioned key
+     *
+     * @param {String} key - object key
+     * @return {String} - versioned base key
+     */
     _extractVersionedBaseKey(key) {
         return key.split(VID_SEP)[0];
     }
 
     /**
-     * Asynchronous filter
-     * @param {Object} entry - constructor params
-     * @param {function} cb - callback
+     * Process bucket entry from the log
+     *
+     * @param {Object} value - log entry object
+     * @return {undefined}
+     */
+    _processBucketEntry(value) {
+        const bucketName = this._getBucketNameFromAttributes(value);
+        const notificationConfiguration
+            = this._getBucketNotificationConfiguration(value);
+        if (notificationConfiguration &&
+            Object.keys(notificationConfiguration).length > 0) {
+            const bnConfig = {
+                bucket: bucketName,
+                notificationConfiguration,
+            };
+            // bucket notification config is available, update node
+            this.bnConfigManager.setConfig(bucketName, bnConfig);
+            return undefined;
+        }
+        // bucket notification conf has been removed, so remove zk node
+        return this.bnConfigManager.removeConfig(bucketName);
```

Comment:
Yes, I would, it keeps the logic simple and easy. Watchers are set for `getChildren` and `getData`, translates to `node children modified`, `node child deleted` and `node updated` (in that order).. I assume when you say `config exists but has no entry` it means a config exists but has `{}` as its value! If the config value is empty for a bucket, we remove it and set it in zookeeper when it is updated (watcher will be triggered and all the manager instances get the update).

Reply:
OK, that works. Make sure this is consistent with the code that adds config too, i.e. it does not add it if the config is empty (otherwise, might still work but you would have an inconsistent handling of empty config, with zookeeper nodes sometimes present, sometimes not, which can lead to subtle bugs)

Reply:
yes, taken care of, https://github.com/scality/backbeat/blob/d8dec0ce11ce311785f16afee13a01c9fb497e88/extensions/notification/NotificationQueuePopulator.js#L108